### DISCUSSION TITLE
Device Service SDK hack to forward a message to a another Service

### DISCUSSION THREAD

**Discussion by waas1919** [edgexfoundry/community#130] [Q&A] [answered]
Good morning

I've been looking at the `Device Service SDK` in C and have a very general question related to this.
I've seen the examples provided and it seems that, by default, once a `device service` gets new data from a "sensor" it pushes it to the `core-data`.
This time, and also based on the answer given by @lenny-intel on this question #128  , I wanted to know what could the best approach for the following scenario:

- I have a device service that generates random numbers (So, no necessarily images or any heavy loads)
- I want to increment the random numbers in another service (let's call it `IncrementService` on the port `59876`)
- After increment the number, I want now to push it to the `core-data`

I see the following options for this, and would like to know which one you advise, or if none, all comments are very welcome.

**[Option 1]**
- Hack into the device service sdk to force pushing the messages to the `IncrementService` on the port `59876`.
- For all data that the `IncrementService` receives, it will increment the number and finally push data to the `core-data`

**[Option 2]**
- Change the `configuration.toml` of the device service where we would change the `Port` to be `IncrementService` on the port `59876`
```
[Clients]
  [Clients.core-data]
  Protocol = 'http'
  Host = 'localhost'
  Port = 59876 # Previously was 59880
``` 
- This case maybe we would receive in our `IncrementService` all the output from this `device service`.
- For all data that the `IncrementService` receives, it will increment the number and finally push data to the `core-data`

**Question:**
As I am asking this, it seems a bit hacky these previous options. I feel I am not really using the best practice here.
Is there any other option, for instance, using the methods from the `Device Service SDK` in `C` where I can force the data to be forwarded into some other port, and not the `core-data`? And then using the `core-data` API, after processing I could simply do a POST to the core-data?
What is your advice?

Thank  you!

**Comment by waas1919** [comment]
Maybe my question is a bit confusing. What I would like to know is that if it's possible to use the `device-sdk-c` to send messages to some other server (could be even a python http server outside of EdgeX)?
In this example:
https://github.com/edgexfoundry/device-sdk-c/blob/main/src/c/examples/counters/device-counter.c#LL126C9-L126C17

This is sending the counter data directly to `core-data`. It is possible to send the data elsewhere?

Thanks

**Comment by lenny-goodell** [comment, marked as answer]
@waas1919 , why wouldn't the device service itself increment the number.

Note that Core Data is optional persistence of Event/Reading data. Device services publish Event/Reading data to the EdgeX MessageBus for other services (App, Rules Engine, Core Data) to consume. 

A custom app service is most appropriate for transforming data (app-Increment) and publishing it back to the EdgeX MessageBus for other services (other App, Rules Engine, Core Data) to consume.
